### PR TITLE
Search hw

### PR DESCRIPTION
**Pull request by BelitskyiAlexandr**
Have some problem. I do not know why, because I checked with the video and tried to debug. This value is nul from the beginning:
image

```

  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/

 :: Spring Boot ::                (v3.3.2)

2024-08-28T20:15:32.678+03:00  INFO 16456 --- [           main] c.e.bookshop.SpringBootIntroApplication  : Starting SpringBootIntroApplication using Java 17.0.2 with PID 16456 (D:\dyst\proj\spring\spring-boot-intro\target\classes started by ADMIN in D:\dyst\proj\spring\spring-boot-intro)
2024-08-28T20:15:32.685+03:00  INFO 16456 --- [           main] c.e.bookshop.SpringBootIntroApplication  : No active profile set, falling back to 1 default profile: "default"
2024-08-28T20:15:35.886+03:00  INFO 16456 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Bootstrapping Spring Data JPA repositories in DEFAULT mode.
2024-08-28T20:15:36.070+03:00  INFO 16456 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Finished Spring Data repository scanning in 170 ms. Found 1 JPA repository interface.
2024-08-28T20:15:38.632+03:00  INFO 16456 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat initialized with port 8080 (http)
2024-08-28T20:15:38.664+03:00  INFO 16456 --- [           main] o.apache.catalina.core.StandardService   : Starting service [Tomcat]
2024-08-28T20:15:38.665+03:00  INFO 16456 --- [           main] o.apache.catalina.core.StandardEngine    : Starting Servlet engine: [Apache Tomcat/10.1.26]
2024-08-28T20:15:38.911+03:00  INFO 16456 --- [           main] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring embedded WebApplicationContext
2024-08-28T20:15:38.914+03:00  INFO 16456 --- [           main] w.s.c.ServletWebServerApplicationContext : Root WebApplicationContext: initialization completed in 6039 ms
2024-08-28T20:15:39.729+03:00  INFO 16456 --- [           main] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Starting...
2024-08-28T20:15:40.642+03:00  INFO 16456 --- [           main] com.zaxxer.hikari.pool.HikariPool        : HikariPool-1 - Added connection com.mysql.cj.jdbc.ConnectionImpl@5b251fb9
2024-08-28T20:15:40.646+03:00  INFO 16456 --- [           main] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Start completed.
2024-08-28T20:15:42.908+03:00  INFO 16456 --- [           main] liquibase.changelog                      : Reading from bookshop_db.databasechangelog
2024-08-28T20:15:43.047+03:00  INFO 16456 --- [           main] liquibase.ui                             : Database is up to date, no changesets to execute
2024-08-28T20:15:43.054+03:00  INFO 16456 --- [           main] liquibase.changelog                      : Reading from bookshop_db.databasechangelog
2024-08-28T20:15:43.070+03:00  INFO 16456 --- [           main] liquibase.util                           : UPDATE SUMMARY
2024-08-28T20:15:43.070+03:00  INFO 16456 --- [           main] liquibase.util                           : Run:                          0
2024-08-28T20:15:43.070+03:00  INFO 16456 --- [           main] liquibase.util                           : Previously run:               1
2024-08-28T20:15:43.070+03:00  INFO 16456 --- [           main] liquibase.util                           : Filtered out:                 0
2024-08-28T20:15:43.070+03:00  INFO 16456 --- [           main] liquibase.util                           : -------------------------------
2024-08-28T20:15:43.070+03:00  INFO 16456 --- [           main] liquibase.util                           : Total change sets:            1
2024-08-28T20:15:43.072+03:00  INFO 16456 --- [           main] liquibase.util                           : Update summary generated
2024-08-28T20:15:43.152+03:00  INFO 16456 --- [           main] liquibase.lockservice                    : Successfully released change log lock
2024-08-28T20:15:43.156+03:00  INFO 16456 --- [           main] liquibase.command                        : Command execution complete
2024-08-28T20:15:43.538+03:00  INFO 16456 --- [           main] o.hibernate.jpa.internal.util.LogHelper  : HHH000204: Processing PersistenceUnitInfo [name: default]
2024-08-28T20:15:43.795+03:00  INFO 16456 --- [           main] org.hibernate.Version                    : HHH000412: Hibernate ORM core version 6.5.2.Final
2024-08-28T20:15:44.068+03:00  INFO 16456 --- [           main] o.h.c.internal.RegionFactoryInitiator    : HHH000026: Second-level cache disabled
2024-08-28T20:15:44.936+03:00  INFO 16456 --- [           main] o.s.o.j.p.SpringPersistenceUnitInfo      : No LoadTimeWeaver setup: ignoring JPA class transformer
2024-08-28T20:15:46.904+03:00  INFO 16456 --- [           main] o.h.e.t.j.p.i.JtaPlatformInitiator       : HHH000489: No JTA platform available (set 'hibernate.transaction.jta.platform' to enable JTA platform integration)
Hibernate: drop table if exists books
Hibernate: create table books (is_deleted bit not null, price decimal(38,2) not null, id bigint not null auto_increment, author varchar(255) not null, cover_image varchar(255), description varchar(255), isbn varchar(255) not null, title varchar(255) not null, primary key (id)) engine=InnoDB
Hibernate: alter table books add constraint UKkibbepcitr0a3cpk3rfr7nihn unique (isbn)
2024-08-28T20:15:47.097+03:00  INFO 16456 --- [           main] j.LocalContainerEntityManagerFactoryBean : Initialized JPA EntityManagerFactory for persistence unit 'default'
2024-08-28T20:15:48.326+03:00  WARN 16456 --- [           main] JpaBaseConfiguration$JpaWebConfiguration : spring.jpa.open-in-view is enabled by default. Therefore, database queries may be performed during view rendering. Explicitly configure spring.jpa.open-in-view to disable this warning
2024-08-28T20:15:49.874+03:00  INFO 16456 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat started on port 8080 (http) with context path '/'
2024-08-28T20:15:49.900+03:00  INFO 16456 --- [           main] c.e.bookshop.SpringBootIntroApplication  : Started SpringBootIntroApplication in 18.363 seconds (process running for 19.462)
2024-08-28T20:16:12.637+03:00  INFO 16456 --- [nio-8080-exec-2] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring DispatcherServlet 'dispatcherServlet'
2024-08-28T20:16:12.638+03:00  INFO 16456 --- [nio-8080-exec-2] o.s.web.servlet.DispatcherServlet        : Initializing Servlet 'dispatcherServlet'
2024-08-28T20:16:12.640+03:00  INFO 16456 --- [nio-8080-exec-2] o.s.web.servlet.DispatcherServlet        : Completed initialization in 2 ms
Hibernate: insert into books (author,cover_image,description,is_deleted,isbn,price,title) values (?,?,?,?,?,?,?)
2024-08-28T20:16:25.056+03:00 ERROR 16456 --- [nio-8080-exec-4] o.a.c.c.C.[.[.[/].[dispatcherServlet]    : Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed: java.lang.NullPointerException: Cannot invoke "com.example.bookshop.repository.SpecificationProviderManager.getSpecificationProvider(String)" because "this.bookSpecificationProviderManager" is null] with root cause

java.lang.NullPointerException: Cannot invoke "com.example.bookshop.repository.SpecificationProviderManager.getSpecificationProvider(String)" because "this.bookSpecificationProviderManager" is null
	at com.example.bookshop.repository.book.BookSpecificationBuilder.build(BookSpecificationBuilder.java:20) ~[classes/:na]
	at com.example.bookshop.service.impl.BookServiceImpl.search(BookServiceImpl.java:59) ~[classes/:na]
	at com.example.bookshop.controller.BookController.search(BookController.java:55) ~[classes/:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:568) ~[na:na]
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:255) ~[spring-web-6.1.11.jar:6.1.11]
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:188) ~[spring-web-6.1.11.jar:6.1.11]
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:118) ~[spring-webmvc-6.1.11.jar:6.1.11]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:926) ~[spring-webmvc-6.1.11.jar:6.1.11]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:831) ~[spring-webmvc-6.1.11.jar:6.1.11]
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87) ~[spring-webmvc-6.1.11.jar:6.1.11]
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1089) ~[spring-webmvc-6.1.11.jar:6.1.11]
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:979) ~[spring-webmvc-6.1.11.jar:6.1.11]
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1014) ~[spring-webmvc-6.1.11.jar:6.1.11]
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:903) ~[spring-webmvc-6.1.11.jar:6.1.11]
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:564) ~[tomcat-embed-core-10.1.26.jar:10.1.26]
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:885) ~[spring-webmvc-6.1.11.jar:6.1.11]
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:658) ~[tomcat-embed-core-10.1.26.jar:10.1.26]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:195) ~[tomcat-embed-core-10.1.26.jar:10.1.26]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.26.jar:10.1.26]
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:51) ~[tomcat-embed-websocket-10.1.26.jar:10.1.26]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[tomcat-embed-core-10.1.26.jar:10.1.26]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.26.jar:10.1.26]
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100) ~[spring-web-6.1.11.jar:6.1.11]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.1.11.jar:6.1.11]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[tomcat-embed-core-10.1.26.jar:10.1.26]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.26.jar:10.1.26]
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93) ~[spring-web-6.1.11.jar:6.1.11]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.1.11.jar:6.1.11]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[tomcat-embed-core-10.1.26.jar:10.1.26]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.26.jar:10.1.26]
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201) ~[spring-web-6.1.11.jar:6.1.11]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.1.11.jar:6.1.11]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[tomcat-embed-core-10.1.26.jar:10.1.26]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.26.jar:10.1.26]
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:167) ~[tomcat-embed-core-10.1.26.jar:10.1.26]
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:90) ~[tomcat-embed-core-10.1.26.jar:10.1.26]
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:483) ~[tomcat-embed-core-10.1.26.jar:10.1.26]
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:115) ~[tomcat-embed-core-10.1.26.jar:10.1.26]
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:93) ~[tomcat-embed-core-10.1.26.jar:10.1.26]
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74) ~[tomcat-embed-core-10.1.26.jar:10.1.26]
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:344) ~[tomcat-embed-core-10.1.26.jar:10.1.26]
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:389) ~[tomcat-embed-core-10.1.26.jar:10.1.26]
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:63) ~[tomcat-embed-core-10.1.26.jar:10.1.26]
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:904) ~[tomcat-embed-core-10.1.26.jar:10.1.26]
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1741) ~[tomcat-embed-core-10.1.26.jar:10.1.26]
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52) ~[tomcat-embed-core-10.1.26.jar:10.1.26]
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1190) ~[tomcat-embed-core-10.1.26.jar:10.1.26]
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659) ~[tomcat-embed-core-10.1.26.jar:10.1.26]
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:63) ~[tomcat-embed-core-10.1.26.jar:10.1.26]
	at java.base/java.lang.Thread.run(Thread.java:833) ~[na:na] 
```

### CODE REVIEW

File: src/main/java/com/example/bookshop/repository/book/spec/AuthorSpecificationProvider.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+@Component
+public class AuthorSpecificationProvider implements SpecificationProvider<Book> {
+    public Specification<Book> getSpecification(String[] authors) {
+        return (root, query, criteriaBuilder) ->
+                root.get("author").in(Arrays.stream(authors).toArray());
+    }
+
+    @Override
+    public String getKey() {
+        return "author";
```
Comment (MentorMateAcademy) [inline src/main/java/com/example/bookshop/repository/book/spec/AuthorSpecificationProvider.java:18]: save string to constants and reuse it

File: src/main/java/com/example/bookshop/repository/book/spec/PriceSpecificationProvider.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    public Specification<Book> getSpecification(String[] prices) {
+        return new Specification<Book>() {
+            @Override
+            public Predicate toPredicate(Root<Book> root, CriteriaQuery<?> query,
+                                         CriteriaBuilder criteriaBuilder) {
+                if (prices != null && prices.length == 2) {
+                    try {
+                        Double minPrice = Double.valueOf(prices[0]);
+                        Double maxPrice = Double.valueOf(prices[1]);
+                        return criteriaBuilder.between(root.get("price"), minPrice, maxPrice);
```
Comment (MentorMateAcademy) [inline src/main/java/com/example/bookshop/repository/book/spec/PriceSpecificationProvider.java:23]: magic numbers

File: src/main/java/com/example/bookshop/repository/book/spec/PriceSpecificationProvider.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+                    } catch (NumberFormatException e) {
+                        // if invalid - true
```
Comment (MentorMateAcademy) [inline src/main/java/com/example/bookshop/repository/book/spec/PriceSpecificationProvider.java:25]: clear comments

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]

**Comment by BelitskyiAlexandr** [comment]
Please, help me with the problem:
https://github.com/BelitskyiAlexandr/spring-boot-intro/pull/5#issue-2492627522

**Review by okuzan — APPROVED** [review]

**Pull request merged**